packages/core/core-flows/src/order/workflows/create-return.ts

```diff
@@ -0,0 +1,305 @@
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Return reason with id ${hasUnExistingReasons.join(
+        ", "
+      )} does not exists.`
+    )
+  }
+
+  if (hasUnusableReasons.length()) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Cannot apply return reason category with id ${hasUnusableReasons.join(
+        ", "
+      )} to order with id ${order_id}.`
+    )
+  }
+}
+
+function prepareShippingMethodData({
+  orderId,
+  inputShippingOption,
+  returnShippingOption,
+}: {
+  orderId: string
+  inputShippingOption: OrderWorkflow.CreateOrderReturnWorkflowInput["return_shipping"]
+  returnShippingOption: ShippingOptionDTO & {
+    calculated_price: { calculated_amount: number }
+  }
+}) {
+  const obj: CreateOrderShippingMethodDTO = {
+    name: returnShippingOption.name,
+    order_id: orderId,
+    shipping_option_id: returnShippingOption.id,
+    amount: 0,
+    data: {},
+    // Computed later in the flow
+    tax_lines: [],
+    adjustments: [],
+  }
+
+  if (isDefined(inputShippingOption.price) && inputShippingOption.price >= 0) {
+    obj.amount = inputShippingOption.price
+  } else {
+    if (returnShippingOption.price_type === "calculated") {
+      // TODO: retrieve calculated price and assign to amount
+    } else {
+      obj.amount = returnShippingOption.calculated_price.calculated_amount
+    }
+  }
+
+  return obj
+}
+
+function validateCustomRefundAmount({
+  order,
+  refundAmount,
+}: {
+  order: Pick<OrderDTO, "item_total">
+  refundAmount?: number
+}) {
+  // validate that the refund prop input is less than order.item_total (item total)
+  // TODO: Probably this amount should be retrieved from the payments linked to the order
+  if (refundAmount && MathBN.gt(refundAmount, order.item_total)) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Refund amount cannot be greater than order total.`
+    )
+  }
+}
+
+function prepareFulfillmentData({
+  order,
+  input,
+}: {
+  order: OrderDTO
+  input: OrderWorkflow.CreateOrderReturnWorkflowInput
+}) {
+  const inputItems = input.items
+  const orderItemsMap = new Map<string, Required<OrderDTO>["items"][0]>(
+    order.items!.map((i) => [i.id, i])
+  )
+  const fulfillmentItems = inputItems.map((i) => {
+    const orderItem = orderItemsMap.get(i.id)!
+    return {
+      line_item_id: i.id,
+      quantity: i.quantity,
+      return_quantity: i.quantity,
+      title: orderItem.variant_title ?? orderItem.title,
+      sku: orderItem.variant_sku || "",
+      barcode: orderItem.variant_barcode || "",
+    } as FulfillmentWorkflow.CreateFulfillmentItemWorkflowDTO
+  })
+
+  // TODO: should we get the stock location from location id and grab the address?
+
+  return {
+    input: {
+      location_id: input.location_id, // TODO: Do we ask the user to provider this, or do we grab the stock location from the fulfillment set of the service zone link to the shipping option id
```

Comment:
I would say default to the fulfillment set location ID, and make it an optional property of the input, so merchants can pass it